[docs] Warn that GradScaler can scale under 1

janeyx99 opened the pull request:
Completes action item 1 in https://github.com/pytorch/pytorch/issues/99640

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101569

janeyx99 committed:
[docs] Warn that GradScaler can scale under 1

[ghstack-poisoned]

janeyx99 commented:
@stas00 @crcrpar Would appreciate your reviews on phrasing!

stas00 commented on docs/source/amp.rst:
```diff
@@ -74,6 +74,15 @@ so they don't flush to zero.
 
+.. note::
+
+  AMP/fp16 may not be for every model! For example, most bf16-pretrained models cannot operate in
+  the fp16 numerical range of max 65k and will cause gradients to overflow instead of underflow. In
+  this case, the scale factor may decrease under 1 as an attempt to bring gradients to a number
+  representable in fp16. While one may expect the scale to always be above 1, our GradScaler does
+  NOT make this guarantee to maintain performance. If you encounter NaNs in your loss or gradients
+  when running with AMP or fp16, verify your model is compatible.
```
```suggestion
  AMP/fp16 may not work for every model! For example, most bf16-pretrained models cannot operate in
  the fp16 numerical range of max 64k and will cause gradients to overflow instead of underflow. In
  this case, the scale factor may decrease under 1 as an attempt to bring gradients to a number
  representable in fp16 dynamic range. While one may expect the scale to always be above 1, our GradScaler does
  NOT make this guarantee to maintain performance. If you encounter NaNs in your loss or gradients
  when running with AMP/fp16, verify your model is compatible.
```

janeyx99 replied:
Thanks for the comments! I thought the max for fp16 is 65.5k or something, no?

stas00 replied:
The maximum representable value is (2−2−10) × 215 = 65504. 

which is 64K (65504/2014)

janeyx99 replied:
Ah, I see the distinction was regarding the K vs k. I'm just going to use the actual number for maximal clarity.

stas00 commented on torch/cuda/amp/grad_scaler.py:
```diff
@@ -396,6 +396,12 @@ def update(self, new_scale=None):
             been invoked for all optimizers used this iteration.
+
+        .. warning::
+            For performance reasons, the scale factor is not guaranteed to be above 1. If the
+            scale falls below 1 and/or you are seeing NaNs in your gradients or loss, something
+            is likely wrong. For example, bf16-pretrained models are often incompatible with
+            AMP/fp16 due to differing numerical ranges.
```
why is this a **performance** reason? I'd call it "numerical stability reasons", no?

janeyx99 replied:
Because adding a check would incur a device sync per step call, and device syncs are expensive

stas00 replied:
oh, I understand what you meant now. 

but if this is not synced how will a user know that it fell below 1? I'm not suggesting to have the overhead, just trying to understand the explanation.

janeyx99 replied:
Oh, like if they check the scale directly as a part of debugging, though I think people would notice the NaNs first

janeyx99 committed:
Update on "[docs] Warn that GradScaler can scale under 1"

Completes action item 1 in https://github.com/pytorch/pytorch/issues/99640




[ghstack-poisoned]

janeyx99 committed:
Update on "[docs] Warn that GradScaler can scale under 1"

Completes action item 1 in https://github.com/pytorch/pytorch/issues/99640




[ghstack-poisoned]

ngimel commented on torch/cuda/amp/grad_scaler.py:
```diff
@@ -396,6 +396,12 @@ def update(self, new_scale=None):
             been invoked for all optimizers used this iteration.
+
+        .. warning::
+            For performance reasons, the scale factor is not guaranteed to be above 1. If the
```
add here that for performance reasons we are not checking scale factor value to avoid synchronization, so it is not guaranteed to be above 1.

janeyx99 committed:
Update on "[docs] Warn that GradScaler can scale under 1"

Completes action item 1 in https://github.com/pytorch/pytorch/issues/99640




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
